internal/db/storage/blob/s3/reader/option_test.go

```diff
@@ -17,465 +17,534 @@
-		           },
-		       }
-		   }(),
-		*/
+	for _, test := range tests {
+		t.Run(test.name, func(tt *testing.T) {
+			defer goleak.VerifyNone(tt, goleakIgnoreOptions...)
+			if test.beforeFunc != nil {
+				test.beforeFunc(test.args)
+			}
+			if test.afterFunc != nil {
+				defer test.afterFunc(test.args)
+			}
+			if test.checkFunc == nil {
+				test.checkFunc = defaultCheckFunc
+			}
+
+			got := WithMaxChunkSize(test.args.size)
+			obj := new(T)
+			got(obj)
+			if err := test.checkFunc(test.want, obj); err != nil {
+				tt.Errorf("error = %v", err)
+			}
+		})
+	}
+}
+
+func TestWithBackoff(t *testing.T) {
+	type T = reader
+	type args struct {
+		enabled bool
+	}
+	type want struct {
+		obj *T
+	}
+	type test struct {
+		name       string
+		args       args
+		want       want
+		checkFunc  func(want, *T) error
+		beforeFunc func(args)
+		afterFunc  func(args)
+	}
+	defaultCheckFunc := func(w want, got *T) error {
+		if !reflect.DeepEqual(got, w.obj) {
+			return errors.Errorf("got = %v, want %v", got, w.obj)
+		}
+		return nil
+	}
+	tests := []test{
+		{
+			name: "set success when backoff enabled is true",
+			args: args{
+				enabled: true,
+			},
+			want: want{
+				obj: &T{
+					backoffEnabled: true,
+				},
+			},
+			checkFunc: defaultCheckFunc,
+		},
+		{
+			name: "set success when backoff enabled is false",
+			want: want{
+				obj: new(T),
+			},
+			checkFunc: defaultCheckFunc,
+		},
 	}
 
 	for _, test := range tests {
 		t.Run(test.name, func(tt *testing.T) {
-			defer goleak.VerifyNone(tt)
+			defer goleak.VerifyNone(tt, goleakIgnoreOptions...)
 			if test.beforeFunc != nil {
 				test.beforeFunc(test.args)
 			}
 			if test.afterFunc != nil {
 				defer test.afterFunc(test.args)
 			}
+			if test.checkFunc == nil {
+				test.checkFunc = defaultCheckFunc
+			}
+
+			got := WithBackoff(test.args.enabled)
+			obj := new(T)
+			got(obj)
+			if err := test.checkFunc(test.want, obj); err != nil {
+				tt.Errorf("error = %v", err)
+			}
 
-			// Uncomment this block if the option returns an error, otherwise delete it
-			/*
-			   if test.checkFunc == nil {
-			       test.checkFunc = defaultCheckFunc
-			   }
+		})
+	}
+}
 
-			   got := WithKey(test.args.key)
-			   obj := new(T)
-			   if err := test.checkFunc(test.want, obj, got(obj)); err != nil {
-			       tt.Errorf("error = %v", err)
-			   }
-			*/
+func TestWithBackoffOpts(t *testing.T) {
+	type T = reader
+	type args struct {
+		opts        []backoff.Option
+		defaultOpts []backoff.Option
+	}
+	type want struct {
+		obj *T
+	}
+	type test struct {
+		name       string
+		args       args
+		want       want
+		checkFunc  func(want, *T) error
+		beforeFunc func(args, *T)
+		afterFunc  func(args, *T)
+	}
+	defaultCheckFunc := func(w want, got *T) error {
+		opts := []cmp.Option{
+			cmp.AllowUnexported(*got),
+			cmp.AllowUnexported(*w.obj),
+			cmp.Comparer(func(want, got []backoff.Option) bool {
+				return reflect.DeepEqual(want, got) || (got != nil && want != nil)
```

Comment:
if you use `DeepEquals` to slice, `DeepEquals` always returns false.
so you should check want, got length, and check element pointer of Options. 🤔

Reply:
according to the document, deepequal can check the elements in the slice. 
https://golang.org/pkg/reflect/#DeepEqual
```
Slice values are deeply equal when all of the following are true: they are both nil or both non-nil, they have the same length, and either they point to the same initial entry of the same underlying array (that is, &x[0] == &y[0]) or their corresponding elements (up to length) are deeply equal. Note that a non-nil empty slice and a nil slice (for example, []byte{} and []byte(nil)) are not deeply equal.
```

Writing this should be sufficient.
```suggestion
				return reflect.DeepEqual(want, got)
```

Reply:
That's right.
And, I have another problem.
`reflect.DeepEqual` will not work well when running test about `set_success_when_opts_is_not_nil_and_backoffOpts_is_not_nil`.
How can I resolve it?
```bash
--- FAIL: TestWithBackoffOpts (0.00s)
    --- FAIL: TestWithBackoffOpts/set_success_when_opts_is_not_nil_and_backoffOpts_is_not_nil (0.00s)
        option_test.go:540: error = got = &{<nil> <nil>   <nil> <nil> false [0xac2060] 0}, want &{<nil> <nil>   <nil> <nil> false [0xac2060] 0}
FAIL
exit status 1
FAIL	github.com/vdaas/vald/internal/db/storage/blob/s3/reader	0.147s
```

Reply:
maybe we need a comparator of `backoff.Option` instead of `[]backoff.Option`
```go
cmp.Comparer(func(want, got backoff.Option) bool
```